File: sdk/python/kflow/service.py
"""ServiceDef and ServiceMode — mirror of Go pkg/kflow/service.go."""
from __future__ import annotations

from enum import IntEnum
from typing import Callable, Optional


class ServiceMode(IntEnum):
    Deployment = 0
    Lambda     = 1


class ServiceDef:
    def __init__(self, name: str) -> None:
        self._name         = name
        self._handler: Optional[Callable] = None
        self._mode         = ServiceMode.Deployment
        self._port         = 8080
        self._min_scale    = 0
        self._max_scale    = 0
        self._ingress_host: Optional[str] = None
        self._timeout      = 30.0  # seconds

    def handler(self, fn: Callable) -> "ServiceDef":
        self._handler = fn
        return self

    def mode(self, mode: ServiceMode) -> "ServiceDef":
        self._mode = mode
        return self

    def port(self, port: int) -> "ServiceDef":
        self._port = port
        return self

    def scale(self, min_: int, max_: int) -> "ServiceDef":
        self._min_scale = min_
        self._max_scale = max_
        return self

    def expose(self, host: str) -> "ServiceDef":
        self._ingress_host = host
        return self

    def timeout(self, seconds: float) -> "ServiceDef":
        self._timeout = seconds
        return self

    def validate(self) -> None:
        """Raises ValueError if any invariant is violated."""
        if self._mode == ServiceMode.Deployment and self._min_scale < 1:
            raise ValueError(
                f"service {self._name!r}: Deployment mode requires min_scale >= 1"
            )


def new_service(name: str) -> ServiceDef:
    """Module-level convenience constructor."""
    return ServiceDef(name)

File: sdk/rust/src/service.rs
use std::time::Duration;

use crate::workflow::HandlerFn;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ServiceMode {
    Deployment = 0,
    Lambda     = 1,
}

impl Default for ServiceMode {
    fn default() -> Self { ServiceMode::Deployment }
}

pub struct ServiceDef {
    pub(crate) name:         String,
    pub(crate) handler:      Option<HandlerFn>,
    pub(crate) mode:         ServiceMode,
    pub(crate) port:         u16,
    pub(crate) min_scale:    u32,
    pub(crate) max_scale:    u32,
    pub(crate) ingress_host: Option<String>,
    pub(crate) timeout:      Duration,
}

impl ServiceDef {
    pub fn new(name: impl Into<String>) -> Self {
        ServiceDef {
            name:         name.into(),
            handler:      None,
            mode:         ServiceMode::Deployment,
            port:         8080,
            min_scale:    0,
            max_scale:    0,
            ingress_host: None,
            timeout:      Duration::from_secs(30),
        }
    }

    pub fn handler(mut self, fn_: HandlerFn) -> Self {
        self.handler = Some(fn_);
        self
    }

    pub fn mode(mut self, mode: ServiceMode) -> Self {
        self.mode = mode;
        self
    }

    pub fn port(mut self, port: u16) -> Self {
        self.port = port;
        self
    }

    pub fn scale(mut self, min: u32, max: u32) -> Self {
        self.min_scale = min;
        self.max_scale = max;
        self
    }

    pub fn expose(mut self, host: impl Into<String>) -> Self {
        self.ingress_host = Some(host.into());
        self
    }

    pub fn timeout(mut self, d: Duration) -> Self {
        self.timeout = d;
        self
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.mode == ServiceMode::Deployment && self.min_scale < 1 {
            return Err(format!(
                "service {:?}: Deployment mode requires min_scale >= 1",
                self.name
            ));
        }
        Ok(())
    }
}

pub fn new_service(name: impl Into<String>) -> ServiceDef {
    ServiceDef::new(name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_service_deployment_requires_min_scale() {
        let svc = new_service("svc").mode(ServiceMode::Deployment).scale(0, 5);
        assert!(svc.validate().unwrap_err().contains("min_scale"));
    }

    #[test]
    fn test_service_lambda_no_constraint() {
        let svc = new_service("svc").mode(ServiceMode::Lambda).scale(0, 5);
        assert!(svc.validate().is_ok());
    }

    #[test]
    fn test_service_deployment_valid() {
        let svc = new_service("svc").mode(ServiceMode::Deployment).scale(1, 5);
        assert!(svc.validate().is_ok());
    }
}
